Authentication configuration system with master password

dakcarto opened the pull request:
Master password-protected storage/retrieval system of authentication configurations, to be used by authentication method plugins to authenticate access to resources, e.g. PKI client certificate presented to SSL server during HTTPS connection.

Companion PR to [associated QEP](https://github.com/qgis/QGIS-Enhancement-Proposals/pull/17). Work sponsored by [Boundless](http://boundlessgeo.com/).

Adds required build/package dependency upon GPL2-licensed [Qt Cryptographic Architecture](http://delta.affinix.com/qca/) (QCA) 2.0.3+, and a run-time dependency upon QCA's OpenSSL plugin (qca-ossl). Latest QCA is **2.1.0**, which builds on all major platforms using CMake and includes all available plugins in the build process.

QCA [source repo browser](http://quickgit.kde.org/?p=qca.git) at KDE.

Mac Homebrew [`qca` formula](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/qca.rb).

dakcarto committed:
[FEATURE] Authentication configuration system with master password

- Main C++ core and gui classes and desktop app integration
- Support for authentication method plugins
- Does not contain any integration with service connections

dakcarto committed:
[auth system] Core data source URI integration

dakcarto committed:
[auth system] basic, pki paths, pkcs12, and identity certificate plugins

dakcarto committed:
[auth system] Data source integration for Postgres and Spit plugin

dakcarto committed:
[auth system] Data source integration for OWS connections

dakcarto committed:
[auth system] Python bindings for core and gui classes

dakcarto committed:
[auth system] Add PKI test data

dakcarto committed:
[auth system] Add to API docs and fix warnings

dakcarto commented:
This supersedes #1838, though the comments about setting up a testing scenario are still valid.

This PR is feature complete with no work left to be done. The only thing currently missing are the unit tests, which were done in Python and are now being migrated over to C++, and will be updated to work with the new authentication method plugin support.

m-kuhn commented:
Wow, nice to see this feature alive.

Is there any documentation available?
The configuration dialog is just a bit too complex for me. Maybe because of limited expertise in the area.

The only thing on the ui which I found to be special so far is the two checkboxes for "show password" in the password and the password confirmation dialog. I think a single one would be just fine (and if it's visible, the confirmation line edit is hidden).

Another question, why are you migrating the tests from python to C++? On my end I almost entirely write tests in python meanwhile, this way the same things are checked plus the python API in addition for free.

dakcarto commented:
Hi @m-kuhn,

> Is there any documentation available?

Yes. I am working on updating the QEP now. We have already written up documentation for the PKI authentication aspect. I recently added support authentication method plugins (C++ only, basically like data providers), so that needs written up a bit as well, for devs looking to add new auth methods.

> The configuration dialog is just a bit too complex for me. Maybe because of limited expertise in the area.

Hmm. I've tried to make those as simple as possible, but then I'm not sure which one you are referring to. We have tested these builds with users for some time now.

> The only thing on the ui which I found to be special so far is the two checkboxes for "show password" in the password and the password confirmation dialog. I think a single one would be just fine (and if it's visible, the confirmation line edit is hidden).

Yes, that makes sense.

> Another question, why are you migrating the tests from python to C++? On my end I almost entirely write tests in python meanwhile, this way the same things are checked plus the python API in addition for free.

I'm not doing away with the Python tests, per se. Just that the biggest possible security hole comes from rogue plugins gaining access to the system. I would like to take the approach of very limited bindings until that can be sorted out, i.e. the tests can be in C++ and only the limited access tested in Python.

As long as PyQGIS plugins use just the core auth config selector widget, and QgsNetworkAccessManager for their network connections, they still get the benefit of the auth system without needing full access to it.

m-kuhn commented:
Sounds good, looking forward

> Hmm. I've tried to make those as simple as possible, but then I'm not sure which one you are referring to. We have tested these builds with users for some time now.

I have been looking at the one in the options dialog. But really, I assume it is a mighty tool and therefore it requires a certain complexity and I am just no yet getting the whole picture.

Just thinking out loud: might it be an idea to offer a simple wizards for setting up basic scenarios for user without bit IT departments and need for certificates?

dakcarto commented:
> Just thinking out loud: might it be an idea to offer a simple wizards for setting up basic scenarios for user without bit IT departments and need for certificates?

The work I was tasked with was to implement PKI. For the normal user, they would just use the *Configurations* tab (auth config selector) in whatever dialog they normally do to configure connections. Currently supported are PostGIS and the OWS connections.

The *Authentication* section in *Options* is more like what you would find in Firefox's options, i.e. editors for settings that are collected during normal use, or so a user/sysadmin can pre-config stuff. Generally, users won't need to go there very often.

There is also a filter setup so that authentication methods that are not implemented for the data provider (or just don't make sense) are hidden in the auth config selector widget. For example, only the Basic auth method is available for PostGIS right now, but PKI can be wired up. Once it is, the newly supported methods will become available to the user.

Conversely, an auth method plugin needs updated to work with various data providers. So, the filter kinda works in reverse, allowing the plugin to tailor how it handles the same method of authentication with a different client setup. Imagine all the ways to authenticate for the OGR provider via the myriad of linked client libs.

Other authentication method plugins could skip the whole configuration storage and just interact with the system's exposure of authentication updating points, e.g. user key cards

m-kuhn commented:
Ah, so I was looking at the wrong dialog to start :)

A short test with the postgres connection dialog works quite well already. **Really nice!!!** :+1: 

I had to go to the Configurations page and setup a new authentication configuration, where some parameters where not immediately clear to me, fortunately they were marked optional so I left them out. Not sure though, I entered a username and a password there and again in the postgres settings dialog. Do they need to be the same?

If I switch the configuration to "No Authentication", the password is dismissed, is that intentionally? Or could a default Configuration be in place of "No Authentication" so I would not even have to go through the "add configuration" procedure?

Sorry for the stupid questions, just trying to keep it as simple as possible for the user who wants the dialogs to "just get out of his way".

dakcarto committed:
[auth system] Fix warnings and add compatibility with QCA 2.0.3

dakcarto commented:
> I had to go to the Configurations page and setup a new authentication configuration, where some parameters where not immediately clear to me, fortunately they were marked optional so I left them out. Not sure though, I entered a username and a password there and again in the postgres settings dialog. Do they need to be the same?

Hmm. Not quite understanding that. Once you have entered an auth config into the system (via the editor in Options-> Authentication or via the auth config selector in a connection dialog) it is available everywhere. When you created a config in the PostGIS dialog it should have shown up in the editor, which is only a table view of the SQLite auth db.

> If I switch the configuration to "No Authentication", the password is dismissed, is that intentionally? Or could a default Configuration be in place of "No Authentication" so I would not even have to go through the "add configuration" procedure?

Maybe that would work. Might be confusing if the user has to manage a default config as well.

You can reuse the same auth config elsewhere in the app. For example, if you set up a username/password that you use for all of your OWS connections via HTTPS, then you can assign that config to every OWS connection setup. Later, if you change your password, you only have to update the auth config, from any dialog where there is an edit button. Then all of your OWS connections are immediately updated.

The authcfg (7-character alphanumeric) token for an auth config can even be used when loading layers via PyQGIS data source strings.

Most of the code is just totally new or bolt-on with very few core integration changes. The rest is really just refactoring and adding the new system over the existing one (username/password) which still works as before.

m-kuhn commented:
Ah, I understand now I think.

I was just being confused by the "Authentication" tab which is also enabled if a "Configuration" has been set up and activated. So I perceived that as a request to enter the credentials again - probably in vain.

Not sure what exactly happens if I save the credentials on the "Authentication" tab and choose a "Configuration". Maybe the old "Authentication" tab could be hidden away or disabled by default (e.g. with a checkable group box).

No further complaints at the moment though, it seems to work all as expected. The only thing is that my tab key doesn't focus widgets in the order I expect them to ;)

dakcarto commented:
> I was just being confused by the "Authentication" tab which is also enabled if a "Configuration" has been set up and activated. So I perceived that as a request to enter the credentials again - probably in vain.

Yes, wanted to leave the old system in place, but it obviously can lead to some confusion. Might be good to have a simple converter routine that migrates a username password over to a new config, then hides the old tab. Personally, I think the users should selectively migrate to the new system.

> Not sure what exactly happens if I save the credentials on the "Authentication" tab and choose a "Configuration". Maybe the old "Authentication" tab could be hidden away or disabled by default (e.g. with a checkable group box).

If an auth config is defined, the old username/password are cleared before saving the dialog. When reopening the dialog, or 'testing' a connection, the tab widget switches to the new configurations tab if a config is in use.

> ... The only thing is that my tab key doesn't focus widgets in the order I expect them to ;)

I will double-check those tab orders. That might get a bit messy with loading auth method edit widgets from factory calls.

nyalldawson commented on src/core/auth/qgsauthmethodregistry.cpp:
```diff
@@ -0,0 +1,392 @@
+      cleanupAuthMethod_t* cleanupFunc = ( cleanupAuthMethod_t* ) cast_to_fptr( myLib.resolve( "cleanupAuthMethod" ) );
+      if ( cleanupFunc )
+        cleanupFunc();
+    }
```
i think this leaks the metadata created in line 140 - needs a 

    delete it->second;

dakcarto replied:
fixed

nyalldawson commented on src/core/auth/qgsauthmanager.cpp:
```diff
@@ -0,0 +1,3291 @@
+
+  query.bindValue( ":id", id );
+
+  
```
nitpicking, but QSSlCertificate should be a reference here

dakcarto replied:
Hmm. Not seeing the issue here. Both the parent function and `QgsAuthCertUtils::shaHexForCert` reference `const QSslCertificate &` values.

nyalldawson replied:
Oops, that note should have been on line 2297. 

There's a few other places with this issue- qgsauthcertutils.cpp:52 and :62, :76:, :86, :596, qgsauthconfig.cpp:275, qgsauthauthoritieseditor.cpp:278, 696, qgsauthcertificateinfo.cpp:201, 777 qgsauthidentitieseditor.cpp:198 qgsauthimportidentitydialog.cpp:418 qgsauthserverseditor.cpp:198 qgsauthtrustedcasdialog.cpp:191

In general it's good practice to always use const references in Q_FOREACH loops, it makes it clear that the loop can't be used to modify the items in the container. Plus, it avoids an unnecessary copy.

nyalldawson commented on src/core/auth/qgsauthcertutils.h:
```diff
@@ -0,0 +1,205 @@
+
+
+    /** SSL Protocol name strings per enum */
+    static const QString getSslProtocolName( QSsl::SslProtocol protocol );
```
these return values don't need to be const, do they?

dakcarto replied:
fixed

nyalldawson commented on src/core/qgsdatasourceuri.h:
```diff
@@ -47,10 +47,10 @@ class CORE_EXPORT QgsDataSourceURI
 
     //! return connection part of URI
-    QString connectionInfo() const;
+    QString connectionInfo( bool expandAuthConfig = true ) const;
```
can you add @param tags for expandAuthConfig? It's probably not obvious what this parameter does if you don't have knowledge of how the auth system works

nyalldawson commented:
@dakcarto I think there needs to be a way to allow sharing of projects which have data sources using the auth system. I'd imagine this is a common requirement (it is at my workplace at least). Allowing users to edit the id of configurations so that they share projects should be sufficient. It probably should have a detailed tooltip or something describing this to users as well.

nyalldawson commented on src/gui/auth/qgsauthauthoritieseditor.cpp:
```diff
@@ -0,0 +1,790 @@
+      QgsAuthImportCertDialog::FileInput );
+  dlg->setWindowModality( Qt::WindowModal );
+  dlg->resize( 400, 250 );
+  if ( dlg->exec() )
```
this leaks dlg if it's not accepted

dakcarto replied:
fixed

nyalldawson commented on src/gui/auth/qgsauthconfigeditor.cpp:
```diff
@@ -0,0 +1,258 @@
+  {
+    mConfigModel->select();
+  }
+}
```
should delete ace here (also at line 198)

dakcarto replied:
fixed

nyalldawson commented on src/gui/auth/qgsauthconfigselect.cpp:
```diff
@@ -0,0 +1,194 @@
+  {
+    setConfigId( ace->configId() );
+  }
+}
```
delete ace here and in line 177

dakcarto replied:
fixed

nyalldawson commented on src/gui/auth/qgsauthimportcertdialog.cpp:
```diff
@@ -0,0 +1,226 @@
+
+  int certssize = certs.size();
+
+  Q_FOREACH ( QSslCertificate cert, certs )
```
change cert to a reference

dakcarto replied:
fixed

nyalldawson commented:
At @dakcarto's request I've given this code a thorough review. I'm not a security or encryption expert, but as far as the code goes overall it's great stuff, nice high quality code and I could only find minor nitpicks with it.

The things which I see need some work are:
- fix Travis failures + add missing doc strings
- add a way for users to manually set the config id so that projects can be shared within a workplace
- also +1 on @m-kuhn's earlier comments re the two show boxes, I don't think that's necessary

dakcarto committed:
[auth system] Fix doc string warnings and comments in PR 2330

dakcarto commented:
Ok. The [QEP has been updated](https://github.com/qgis/QGIS-Enhancement-Proposals/pull/17).

Only things left are ...

* Finishing the port of the Python unit tests to C++
* Add the `authcfg` ID editing, as per @nyalldawson suggestion
* Doc strings are OK, but would like to add to some of the parameter descriptions in non-auth classes

luipir commented:
Hi @dakcarto wonderful PR!
Days ago a friend asked me a way to share a project hiding credentials to acces some OWSs. Your PR would solve this, but I need to figure out if the PR would cover this simple Use Case.
I imagine that my friend, to leave the OWSs pwd encrypted, have to share the project packed with a custom auth.db. This friend have to:
1) create a new auth.db
2) set a master pwd
3) create the config hash
4) store/crypt the pwds when adding layer to the project
5) save the project
6) export auth.db
7) pack together .prj and auth.db and share the master pwd

My questions are:
1) is this procedure correct?
2) I didn't see any export/import function of the auth.db
only reset here:
https://github.com/dakcarto/QGIS-Enhancement-Proposals/blob/auth-system/qep-14-authentication-system.rst
and here:
https://github.com/dakcarto/QGIS-Enhancement-Proposals/blob/auth-system/extras/auth-system/pkiuser.rst
(then the problem of overwrite erroneously a saved auth.db)
3) As I understood, after entered master pwd, qgis or plugins have access only to hashed pwd and it's not possibile to browse the auth.db that remain encrypted. If is it correct, this model would work for my friend's use case.

thank you for this contribution

nyalldawson commented on src/gui/auth/qgsauthimportidentitydialog.cpp:
```diff
@@ -0,0 +1,468 @@
+/***************************************************************************
+    qgsauthimportidentitydialog.h
```
.cpp

dakcarto replied:
fixed

nyalldawson commented:
@dakcarto another minor thing, but do you think it's necessary to have the separate menu item for Authentication?

Given that all it does is open the options at the authentication page, it doesn't seem necessary to me. I can't imagine users will be needing to change this configuration so often that selecting options->authentication is a burden.

strk commented:
About python vs. C++ tests, I'd add that it's easier to spot memory leaks with a C++ test than it is with a python test (correct me if I'm wrong).

nyalldawson commented:
@strk true, given that it's currently not possible to run the python tests using lsan. I also find it easier to debug c++ tests as I've never been able reliably run the python tests under gdb (if anyone has suggestions for this, I'd love to hear...)

dakcarto commented:
@strk Good point. Another reason for strictly C++ tests to start with is security. As an initial release it may be warranted to have _almost no bindings_ to the auth system and require Python plugin developers to use the limited bindings for selecting auth configs and open up the `QgsNetworkAccessManager` to directly work with `authcfg` ID tokens.

This would allow plugin devs to use the auth system while possibly limiting security concerns. They can select a config (which obfuscates credentials) and save it's `authcfg` ID to their plugin's settings. Then when making a connection in their code, they must use the binding to `QgsNetworkAccessManager` for the network resource connection, instead of say Python libs like `requests` or `httplib2`, which would require the `authcfg` to be expanded and expose the credentials to the plugin to be passed to the Python lib.

There may be other approaches to solving this issue. See [QEP section on security](https://github.com/dakcarto/QGIS-Enhancement-Proposals/blob/auth-system/qep-14-authentication-system.rst#5-security-considerations) where I mention using code signing certificates issue via the QGIS project.

Of course, the upshot of this means only those limited bindings could have Python unit tests, and the rest needs to be in C++. I will commit my current Python tests once I have converted them to C++.

dakcarto commented:
@nyalldawson wrote:
> ... another minor thing, but do you think it's necessary to have the separate menu item for Authentication?

Probably not, as most auth system interaction can be accomplished directly through the small `QgsAuthConfigSelect` selector widget (in core or plugin GUIs). I needed to have that menu item for the implementation I did so there was a simple access point to the system for configuring PKI that could be added to user docs.

Simple enough though to just say "Open `Settings -> Options` and go to `Authentication` section for additional settings" in user docs.

Similarly, the PKI-related editors in the `Authentication` section can be isolated in a separate dialog that is accessed via a `Manage Certificates` button (like in Firefox). So there is clear separation between general auth configs and PKI stuff.

dakcarto commented:
@luipir wrote:
> Days ago a friend asked me a way to share a project hiding credentials to acces some OWSs. Your PR would solve this, but I need to figure out if the PR would cover this simple Use Case.
> I imagine that my friend, to leave the OWSs pwd encrypted, have to share the project packed with a custom auth.db. This friend have to:
> 1) create a new auth.db
> 2) set a master pwd
> 3) create the config hash
> 4) store/crypt the pwds when adding layer to the project
> 5) save the project
> 6) export auth.db
> 7) pack together .qgs and auth.db and share the master pwd
> 
> My questions are:
> 1) is this procedure correct?

Yes, though you can achieve the same result by resetting the password and choosing to keep a backup of your current `qgis-auth.db`, which creates a new database encrypted with the new master password, _but_ all the `authcfg` IDs remain the same.

Then you can distribute the project and new db. Before running your QGIS again, swap out the reset db with the saved backup db. No issue with corruption as the old database was not actually touched (just renamed). This is so it can be reinstated if there was an error during password resetting.

> 2) I didn't see any export/import function of the auth.db (ok it's in the todo list)

This sounds like a great feature to add. Choose "Export authentication database" (maybe from the config editor in `Options -> Authentication`).

User could be presented with a choice:

* Export whole database
* Export only configurations associated with current project
* Choose what to export
* (Maybe something else here)

Then the items to export are sent to an export file (not a database), so that it requires the user import them. Since the sensitive data is (_should_) be encrypted with a temp password, an import function would ask for it. Then the import would re-encrypt the credentials with the user's master password into the user's current auth db, prompting if there is any conflict with existing `authcfg` IDs.

While this does mean sharing auth credentials, it is a means of sharing. Likewise, the ability to change the `authcfg` ID within an existing db, as well as selecting replacement auth configs in the **Handles Bad Layers** dialog, should handle most sharing situations.

Another nice way to possibly solve sharing issues is by finishing the `Resource URL` functionality of an auth config. When this is implemented (a bit tricky), **_any_** existing config that is found to match the requested resource URL, and authentication method to use, can be used. Then users just have to have a configuration that _matches_ the data source location. No sharing of auth configs would be needed since no `authcfg` ID is actually associated with the layer.

> 3) As I understood, after entered master pwd, qgis or plugins have access only to hashed pwd and it's not possibile to browse the auth.db that remain encrypted. If is it correct, this model would work for my friend's use case.

The API to the auth system currently does allow access to other credentials _once_ the master password has been set, or the user would be repeatedly requested to enter it. The qgis-auth.db can not be browsed otherwise since the stored data is encrypted, though you can open it up in your favorite SQLite editor to work on it at the database/table/record level.

It might be good to have repeated master password requests, or some form of confirmation, _per_ auth config accessed (as a non-default user preference) to ensure all access is locked down unless authorized. However, this is not like other auth systems and will no doubt annoy users.

nyalldawson commented:
Any objections if I merge this? Will be nice to have in 2.12, and I think it's safe to do given that it's already been tested with the boundless builds, and since it's effectively a parallel implementation.

dakcarto committed:
[auth] Update python tests (keep disabled); add auth config tests

- Add comparison operators to QgsAuthMethodConfig

dakcarto committed:
[auth] Update auth config tests; strip passphrase from QgsPkiBundle

dakcarto committed:
[auth] Strip translated display name from auth method plugins

dakcarto committed:
[auth] Remove Authentication action from Settings menu

dakcarto committed:
[auth] Redesign Authentication section of Options

- Separate management options from auth config table
- Add standalone certificate manager
- Add installed auth method plugin dialog, indicating usage per plugin

dakcarto committed:
[auth] Add authcfg ID edit widget and support for changing the ID

dakcarto commented:
@nyalldawson wrote:
> Thanks!

No problem. Well, it was a little bit trickier than anticipated making the ID editable. Works very well now though.

I am working on another widget that can be used for the Handle Bad Layers and the Layers Properties dialogs. That should work for most issues of sharing and when users open projects with missing authentication IDs. I've also added feedback in the selector widget when it tries to load a config ID that is missing.

I should have those bits done today, then it mostly work on finishing the unit test migration to C++.

However, I would like to *merge this now*, since it is in a very good state for doing so. **Does anyone have objections with me merging this?**

luipir commented:
@dakcarto +1 to merge for 2.12

nyalldawson commented:
@dakcarto merged. Thanks Larry for all the work and polish you've put into this. It's fantastic stuff!

blazek commented:
Great work, +1 for merge from me, even post merge.

When I now launch QGIS, I get (BTW, impossible to copy paste the error):
https://www.google.com/jsapi
Unable To Get Local Issuer Certificate...
Certificate Untrusted: ...

As I am impatient trying to set it up, I simply check "Save SSL server exception" + "Save & Ignore", but the next time I launch QGIS, I get the same error.  Am I doing something wrong? 

Developing and testing, I am continuously starting QGIS... Is it chance to get it fixed quickly?

blazek commented:
I got over setting "Peer verification" to "Do not verify peer certs". Having checked "Ignore errors": "Unable to Get Local Issuer Certificate" and "Certificate Untrusted" was not enough.

I am curious where the request for www.google.com comes from.

m-kuhn commented:
For the rss feed javascript api on the what's new page.

https://github.com/qgis/QGIS/blob/master/doc/whatsnew.html#L87

dakcarto commented:
@blazek wrote:
> When I now launch QGIS, I get (BTW, impossible to copy paste the error):
> https://www.google.com/jsapi
> Unable To Get Local Issuer Certificate...
> Certificate Untrusted: ...

That is a recent bug that I have not be able to understand why it occurs (started happening after the proxy support, i.e. [using `QgsNetworkAccessManager`](https://github.com/qgis/QGIS/commit/8b9e0b65d63fb6de8a12f588cc3ff3c09e92ed01), was added to the new welcome page).

Setting the SSL cert exception is a workaround, but should not be needed. It appears that no trusted Certificate Authorities are being added to the `QgsWebView` requests. This is very odd, especially because if you inspect the certificate chain and trusted certs from within the SSL error dialog everything is there. I even tried adding the specific cert authorities to QGIS's new cert manager, with no luck.

Next I tried commenting out [the `QNetworkRequest` hack in `QgsNetworkAccessManager` for HTTPS requests](https://github.com/qgis/QGIS/blob/master/src/core/qgsnetworkaccessmanager.cpp#L172-L195), but that didn't work either.

It seems that the QWebPage requests are maybe in a different thread and the SSL errors, as handled by the `QgisApp`, are just a response to a signal, but not being honored in the calling thread. Still it doesn't explain why the CAs are not being passed to the `QgsWebView` to begin with.

> I got over setting "Peer verification" to "Do not verify peer certs". Having checked "Ignore errors": "Unable to Get Local Issuer Certificate" and "Certificate Untrusted" was not enough.

Hmm. On my Mac, just leaving the "Ignore errors" worked for me.

Any ideas on what might be causing this?
